Python 3.15 support

[heads up] AssertionError in test_dont_write_bytecode and leaking deprecation warnings with Python 3.15.0b1

In Fedora Linux we are building packages with Python 3.15. While pytest itself builds fine and we have successfully used it for many dependent packages, its own test suite doesn't pass the run.

There are two issues with Python 3.15.0b1, one actual AssertionError and 38 deprecation warnings leaking from ptyprocess, treated as errors.
We have pytest 9.0.3, the first failure is reproducible on current main with `tox -e py315`.
The deprecation warnings appear only in our downstream build, I couldn't reproduce with the current main nor 9.0.3 tag.

1.
```
_________________ TestRewriteOnImport.test_dont_write_bytecode _________________
[gw3] linux -- Python 3.15.0 /usr/bin/python3

self = <test_assertrewrite.TestRewriteOnImport object at 0x7fa98bbffdf0>
pytester = <Pytester PosixPath('/tmp/pytest-of-mockbuild/pytest-0/popen-gw3/test_dont_write_bytecode0')>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fa988ff4750>

    def test_dont_write_bytecode(self, pytester: Pytester, monkeypatch) -> None:
        monkeypatch.delenv("PYTHONPYCACHEPREFIX", raising=False)
    
        pytester.makepyfile(
            """
            import os
            def test_no_bytecode():
                assert "__pycache__" in __cached__
                assert not os.path.exists(__cached__)
                assert not os.path.exists(os.path.dirname(__cached__))"""
        )
        monkeypatch.setenv("PYTHONDONTWRITEBYTECODE", "1")
>       assert pytester.runpytest_subprocess().ret == 0
E       AssertionError: assert <ExitCode.TESTS_FAILED: 1> == 0
E        +  where <ExitCode.TESTS_FAILED: 1> = <RunResult ret=1 len(stdout.lines)=19 len(stderr.lines)=0 duration=0.30s>.ret
E        +    where <RunResult ret=1 len(stdout.lines)=19 len(stderr.lines)=0 duration=0.30s> = runpytest_subprocess()
E        +      where runpytest_subprocess = <Pytester PosixPath('/tmp/pytest-of-mockbuild/pytest-0/popen-gw3/test_dont_write_bytecode0')>.runpytest_subprocess

/builddir/build/BUILD/pytest-9.0.3-build/pytest-9.0.3/testing/test_assertrewrite.py:1090: AssertionError
----------------------------- Captured stdout call -----------------------------
running: /usr/bin/python3 -mpytest --basetemp=/tmp/pytest-of-mockbuild/pytest-0/popen-gw3/test_dont_write_bytecode0/runpytest-0
     in: /tmp/pytest-of-mockbuild/pytest-0/popen-gw3/test_dont_write_bytecode0
============================= test session starts ==============================
platform linux -- Python 3.15.0b1, pytest-9.0.3, pluggy-1.6.0
rootdir: /tmp/pytest-of-mockbuild/pytest-0/popen-gw3/test_dont_write_bytecode0
collected 1 item

test_dont_write_bytecode.py F                                            [100%]

=================================== FAILURES ===================================
_______________________________ test_no_bytecode _______________________________

    def test_no_bytecode():
>       assert "__pycache__" in __cached__
                                ^^^^^^^^^^
E       NameError: name '__cached__' is not defined

test_dont_write_bytecode.py:3: NameError
```

```
_______________________ TestTrialUnittest.test_trial_pdb _______________________
[gw2] linux -- Python 3.15.0 /usr/bin/python3

self = <test_unittest.TestTrialUnittest object at 0x7f126977c2b0>
pytester = <Pytester PosixPath('/tmp/pytest-of-mockbuild/pytest-0/popen-gw2/test_trial_pdb0')>

    def test_trial_pdb(self, pytester: Pytester) -> None:
        p = pytester.makepyfile(
            """
            from twisted.trial import unittest
            import pytest
            class TC(unittest.TestCase):
                def test_hello(self):
                    assert 0, "hellopdb"
        """
        )
>       child = pytester.spawn_pytest(str(p))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

/builddir/build/BUILD/pytest-9.0.3-build/pytest-9.0.3/testing/test_unittest.py:617: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/builddir/build/BUILD/pytest-9.0.3-build/BUILDROOT/usr/lib/python3.15/site-packages/_pytest/pytester.py:1531: in spawn_pytest
    return self.spawn(cmd, expect_timeout=expect_timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/builddir/build/BUILD/pytest-9.0.3-build/BUILDROOT/usr/lib/python3.15/site-packages/_pytest/pytester.py:1545: in spawn
    child = pexpect.spawn(cmd, logfile=logfile, timeout=expect_timeout)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.15/site-packages/pexpect/pty_spawn.py:205: in __init__
    self._spawn(command, args, preexec_fn, dimensions)
/usr/lib/python3.15/site-packages/pexpect/pty_spawn.py:303: in _spawn
    self.ptyproc = self._spawnpty(self.args, env=self.env,
/usr/lib/python3.15/site-packages/pexpect/pty_spawn.py:315: in _spawnpty
    return ptyprocess.PtyProcess.spawn(args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.15/site-packages/ptyprocess/ptyprocess.py:230: in spawn
    pid, fd = pty.fork()
              ^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def fork():
        """fork() -> (pid, master_fd)
        Fork and make the child a session leader with a controlling terminal."""
    
        try:
>           pid, fd = os.forkpty()
                      ^^^^^^^^^^^^
E           DeprecationWarning: This process (pid=1140) is multi-threaded, use of forkpty() may lead to deadlocks in the child.
```

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -65,6 +65,7 @@ jobs:
           "windows-py312",
           "windows-py313",
           "windows-py314",
+          "windows-py315",
 
           "ubuntu-py310-unittest-asynctest",
           "ubuntu-py310-unittest-twisted24",
@@ -77,12 +78,14 @@ jobs:
           "ubuntu-py312",
           "ubuntu-py313-pexpect",
           "ubuntu-py314",
+          "ubuntu-py315",
           "ubuntu-pypy3-xdist",
 
           "macos-py310",
           "macos-py312",
           "macos-py313",
           "macos-py314",
+          "macos-py315",
 
           "doctesting",
           "plugins",
@@ -140,6 +143,11 @@ jobs:
             os: windows-latest
             tox_env: "py314"
             use_coverage: true
+
+          - name: "windows-py315"
+            python: "3.15"
+            os: windows-latest
+            tox_env: "py315"
 
           # Use separate jobs for different unittest flavors (twisted, asynctest) to ensure proper coverage.
           - name: "ubuntu-py310-unittest-asynctest"
@@ -208,6 +216,12 @@ jobs:
             tox_env: "py314"
             use_coverage: true
 
+          - name: "ubuntu-py315"
+            python: "3.15"
+            os: ubuntu-latest
+            tox_env: "py315"
+            use_coverage: true
+
           - name: "ubuntu-pypy3-xdist"
             python: "pypy-3.10"
             os: ubuntu-latest
@@ -235,6 +249,11 @@ jobs:
             python: "3.14"
             os: macos-latest
             tox_env: "py314-xdist"
+
+          - name: "macos-py315"
+            python: "3.15"
+            os: macos-latest
+            tox_env: "py315-xdist"
 
           - name: "plugins"
             python: "3.12"

--- a/changelog/14524.improvement.rst
+++ b/changelog/14524.improvement.rst
@@ -1,0 +1,1 @@
+Add official Python 3.15 support.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -415,6 +415,8 @@ filterwarnings = [
     # https://github.com/pytest-dev/pytest/issues/2366
     # https://github.com/pytest-dev/pytest/pull/13057
     'default::pytest.PytestFDWarning',
+    # https://github.com/pexpect/pexpect/issues/827
+    'ignore:.*use of forkpty\(\) may lead to deadlocks in the child:DeprecationWarning:ptyprocess.*',
 ]
 pytester_example_dir = "testing/example_scripts"
 markers = [

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -38,6 +38,7 @@ classifiers = [
     "Programming Language :: Python :: 3.12",
     "Programming Language :: Python :: 3.13",
     "Programming Language :: Python :: 3.14",
+    "Programming Language :: Python :: 3.15",
     "Topic :: Software Development :: Libraries",
     "Topic :: Software Development :: Testing",
     "Topic :: Utilities",
@@ -353,7 +354,7 @@ ignore = "W009"
 
 [tool.pyproject-fmt]
 indent = 4
-max_supported_python = "3.14"
+max_supported_python = "3.15"
 
 [tool.pytest]
 minversion = "2.0"

--- a/testing/test_assertrewrite.py
+++ b/testing/test_assertrewrite.py
@@ -1082,9 +1082,10 @@ def test_rewritten():
             """
             import os
             def test_no_bytecode():
-                assert "__pycache__" in __cached__
-                assert not os.path.exists(__cached__)
-                assert not os.path.exists(os.path.dirname(__cached__))"""
+                assert "__pycache__" in __spec__.cached
+                assert not os.path.exists(__spec__.cached)
+                assert not os.path.exists(os.path.dirname(__spec__.cached))
+            """
         )
         monkeypatch.setenv("PYTHONDONTWRITEBYTECODE", "1")
         assert pytester.runpytest_subprocess().ret == 0

--- a/tox.ini
+++ b/tox.ini
@@ -8,6 +8,7 @@ envlist =
     py312
     py313
     py314
+    py315
     pypy3
     py310-{pexpect,xdist,twisted24,twisted25,asynctest,numpy,pluggymain,pylib}
     doctesting

